src/dft/descriptor_config_helper.hpp

```diff
@@ -0,0 +1,319 @@
+        param_type_helper_t<real_helper_t<prec>, config_param::INPUT_STRIDES>& set_val) {
+        int rank = vals.rank;
+        if (set_val == nullptr) {
+            throw mkl::invalid_argument("DFT", "set_value", "Given nullptr.");
+        }
+        std::copy(set_val, set_val + vals.rank, vals.input_strides.begin());
+    }
+};
+
+template <precision prec, domain dom>
+struct set_value_helper<config_param::OUTPUT_STRIDES, prec, dom> {
+    static void set_val(
+        dft_values<prec, dom>& vals,
+        param_type_helper_t<real_helper_t<prec>, config_param::OUTPUT_STRIDES>& set_val) {
+        int rank = vals.rank;
+        if (set_val == nullptr) {
+            throw mkl::invalid_argument("DFT", "set_value", "Given nullptr.");
+        }
+        std::copy(set_val, set_val + vals.rank, vals.output_strides.begin());
+    }
+};
+
+template <precision prec, domain dom>
+struct set_value_helper<config_param::FWD_DISTANCE, prec, dom> {
+    static void set_val(
+        dft_values<prec, dom>& vals,
+        param_type_helper_t<real_helper_t<prec>, config_param::FWD_DISTANCE>& set_val) {
+        vals.fwd_dist = set_val;
+    }
+};
+
+template <precision prec, domain dom>
+struct set_value_helper<config_param::BWD_DISTANCE, prec, dom> {
+    static void set_val(
+        dft_values<prec, dom>& vals,
+        param_type_helper_t<real_helper_t<prec>, config_param::BWD_DISTANCE>& set_val) {
+        vals.bwd_dist = set_val;
+    }
+};
+
+template <precision prec, domain dom>
+struct set_value_helper<config_param::WORKSPACE, prec, dom> {
+    static void set_val(
+        dft_values<prec, dom>& vals,
+        param_type_helper_t<real_helper_t<prec>, config_param::WORKSPACE>& set_val) {
+        if (set_val == config_value::ALLOW || set_val == config_value::AVOID) {
+            vals.workspace = set_val;
+        }
+        else {
+            throw mkl::invalid_argument("DFT", "set_value", "Workspace must be allow or avoid.");
+        }
+    }
+};
+
+template <precision prec, domain dom>
+struct set_value_helper<config_param::ORDERING, prec, dom> {
+    static void set_val(dft_values<prec, dom>& vals,
+                        param_type_helper_t<real_helper_t<prec>, config_param::ORDERING>& set_val) {
+        if (set_val == config_value::ORDERED || set_val == config_value::BACKWARD_SCRAMBLED) {
+            vals.ordering = set_val;
+        }
+        else {
+            throw mkl::invalid_argument("DFT", "set_value",
+                                        "Ordering must be ordered or backwards scrambled.");
+        }
+    }
+};
+
+template <precision prec, domain dom>
+struct set_value_helper<config_param::TRANSPOSE, prec, dom> {
+    static void set_val(
+        dft_values<prec, dom>& vals,
+        param_type_helper_t<real_helper_t<prec>, config_param::TRANSPOSE>& set_val) {
+        vals.transpose = set_val;
+    }
+};
+
+template <precision prec, domain dom>
+struct set_value_helper<config_param::PACKED_FORMAT, prec, dom> {
```

Comment:
What is the advantage of having the set_value_helper struct instead of just set_val as a template_function?

Reply:
The `dft_values` class now takes the precision and domain as template arguments. Whilst the need for domain argument is debatable, the precision argument is required since it determines the type of the forward scale. 

The consequence of this is that set value requires partial specialization. This cannot be achieved with functions alone.